processing/src/main/java/io/druid/query/QueryMetrics.java

```diff
@@ -0,0 +1,198 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import com.metamx.emitter.service.ServiceEmitter;
+
+import java.util.Map;
+
+/**
+ * Abstraction wrapping {@link com.metamx.emitter.service.ServiceMetricEvent.Builder} and allowing to control what
+ * metrics are actually emitted, what dimensions do they have, etc.
+ *
+ *
+ * Goals of QueryMetrics
+ * ---------------------
+ *  1. Skipping or partial filtering of particular dimensions or metrics entirely. Implementation could leave the body
+ *  of the corresponding method empty, or implement random filtering like:
+ *  public void cpuTime(long timeNs)
+ *  {
+ *    if (ThreadLocalRandom.current().nextDouble() < 0.1) {
+ *      super.cpuTime(timeNs);
+ *    }
+ *  }
+ *
+ *  2. Ability to add new dimensions and metrics, possibly expensive to compute, or expensive to process (long string
+ *  values, high cardinality, etc.) and not to affect existing Druid installations, by skipping (see 1.) those
+ *  dimensions and metrics entirely in the default QueryMetrics implementations. Users who need those expensive
+ *  dimensions and metrics, could explicitly emit them in their own QueryMetrics.
+ *
+ *  3. Control over the time unit, in which time metrics are emitted. By default (see {@link DefaultQueryMetrics} and
+ *  it's subclasses) it's milliseconds, but if queries are fast, it could be not precise enough.
+ *
+ *  4. Control over the dimension and metric names.
+ *
+ *
+ * Types of methods in this interface
+ * ----------------------------------
+ *  1. Methods, pulling some dimensions from the query object. These methods are used to populate the metric before the
+ *  query is run. These methods accept a single `QueryType query` parameter. {@link #query(Query)} calls all methods
+ *  of this type, hence pulling all available information from the query object as dimensions.
+ *  2. Methods for setting dimensions, which become known in the process of the query execution or after the query is
+ *  completed.
+ *  3. Methods to register metrics to be emitted later in bulk via {@link #emit(ServiceEmitter)}. These methods
+ *  return this QueryMetrics object back for chaining.
+ *
+ *
+ * Implementors expectations
+ * -------------------------
+ * QueryMetrics is expected to be changed often, in every Druid release. Users who create their custom implementations
+ * of QueryMetrics should be ready to fix the code of their QueryMetrics (implement new methods) when they update Druid.
+ * Broken builds of custom extensions, containing custom QueryMetrics is the way to notify users that Druid core "wants"
+ * to emit new dimension or metric, and the user handles them manually: if the new dimension or metric is useful and not
+ * very expensive to process and store then emit, skip (see above Goals, 1.) otherwise.
+ *
+ * <p>QueryMetrics is designed for use from a single thread, implementations shouldn't care about thread-safety.
```

Comment:
I admit this is getting into the realm where my JVM memory consistency guarantees gets a little funny

https://github.com/metamx/emitter/blob/master/src/main/java/com/metamx/emitter/core/HttpPostEmitter.java#L174

Here a queue is being added to for being operated on by a runnable in a parallel thread.

The object value consistency is kept because the emit method is in the same thread as the serialization at https://github.com/metamx/emitter/blob/master/src/main/java/com/metamx/emitter/core/HttpPostEmitter.java#L156 

the emitting is then done in a separate thread as per
https://github.com/metamx/emitter/blob/master/src/main/java/com/metamx/emitter/core/HttpPostEmitter.java#L259

If the emitting object were passed instead of the byte array, I think the synchronization would be enough to keep memory consistency between the thread calling `emit` and the thread executing `EmittingRunnable`.

Such a thing could be very easy to screw up though.